### PR TITLE
Fixed Project API tests

### PR DESCRIPTION
**Pull request by ramakantar**
- Added Project API tests
- Added Project API example

### CODE REVIEW

File: inventories_test.go [hunk 1]
```diff
@@ -531,15 +534,14 @@ func TestGetInventory(t *testing.T) {
...
-	}
 
-	if !reflect.DeepEqual(result, expectGetInventoryResponse) {
-		log.Fatalf("GetInventory by ID resp not as expected, expected: %v, resp result: %v", expectGetInventoryResponse, result)
+	if err != nil {
+		t.Fatalf("GetInventory by ID err: %s", err)
+	} else if !reflect.DeepEqual(result, expectGetInventoryResponse) {
+		t.Fatalf("GetInventory by ID resp not as expected, expected: %v, resp result: %v", expectGetInventoryResponse, result)
+	} else {
+		log.Println("GetInventory by ID passed!")
```
Comment (Colstuwjx) [inline inventories_test.go]: Please keep same code style with other test cases.

File: projects.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// ListProjectsResponse represents `ListProjects` endpoint response.
+type ListProjectsResponse struct {
+	Pagination
+	Results []*Project `json:"results"`
+}
+
+// ListProjects shows list of awx projects.
+func (i *ProjectService) ListProjects(params map[string]string) ([]*Project, *ListProjectsResponse, error) {
```
Comment (Colstuwjx) [inline projects.go]: `i` is not clear, should be `p` for go style.

**Comment by ramakantar** [comment]
Fixed the test log types

File: projects_test.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	)
+
+	awx := NewAWX(testAwxHost, testAwxUserName, testAwxPasswd, nil)
+	result, err := awx.ProjectService.CreateProject(map[string]interface{}{
+		"name":        "TestProject",
+		"description": "Test project",
+	}, map[string]string{})
+
+	if err != nil {
+		t.Fatalf("CreateInventory err: %s", err)
```
Comment (Colstuwjx) [inline projects_test.go]: typo here

File: job_test.go [hunk 4]
```diff
@@ -530,13 +533,14 @@ func TestGetJobEvents(t *testing.T) {
...
-	}
 
-	if !reflect.DeepEqual(result, expectJobEventsResponse) {
-		log.Fatalf("GetJobEvents resp not as expected, expected: %v, resp result: %v", expectJobEventsResponse, result)
+	if err != nil {
+		t.Fatalf("GetJobEvents err: %s", err)
+	} else if !reflect.DeepEqual(result, expectJobEventsResponse) {
+		t.Logf("expected: %v", expectJobEventsResponse)
+		t.Logf("result: %v", result)
+		t.Fatal("GetJobEvents resp not as expected, expected")
```
Comment (Colstuwjx) [inline job_test.go]: Correct the message

File: job_test.go [hunk 5]
```diff
@@ -454,15 +456,16 @@ func TestGetHostSummaries(t *testing.T) {
...
-	}
 
-	if !reflect.DeepEqual(result, expectHostSummariesResponse) {
-		log.Fatalf("GetHostSummaries resp not as expected, expected: %v, resp result: %v", expectHostSummariesResponse, result)
+	if err != nil {
+		t.Fatalf("GetHostSummaries err: %s", err)
+	} else if !reflect.DeepEqual(result, expectHostSummariesResponse) {
+		t.Logf("expected: %v", expectHostSummariesResponse)
+		t.Logf("result: %v", result)
+		t.Fatal("GetHostSummaries resp not as expected, expected")
```
Comment (Colstuwjx) [inline job_test.go]: Correct the message

**Comment by ramakantar** [comment]
Fixed it.

File: job_test.go [hunk 6]
```diff
@@ -539,7 +539,7 @@ func TestGetJobEvents(t *testing.T) {
 	} else if !reflect.DeepEqual(result, expectJobEventsResponse) {
 		t.Logf("expected: %v", expectJobEventsResponse)
 		t.Logf("result: %v", result)
-		t.Fatal("GetJobEvents resp not as expected, expected")
+		t.Fatal("GetJobEvents resp	 not as expected")
```
Comment (Colstuwjx) [inline job_test.go]: Too many spaces here.

File: types.go [hunk 7]
```diff
@@ -538,3 +553,20 @@ type JobEvent struct {
 	EndLine   int         `json:"end_line"`
 	Verbosity int         `json:"verbosity"`
 }
+
+// User represents an user
+type User struct {
```
Comment (Colstuwjx) [inline types.go]: In my local test, I got user response from awx rest api `/api/v2/users/1` below:

```
{
    "id": 1,
    "type": "user",
    "url": "/api/v2/users/1/",
    "related": {
        "named_url": "/api/v2/users/admin/",
        "admin_of_organizations": "/api/v2/users/1/admin_of_organizations/",
        "authorized_tokens": "/api/v2/users/1/authorized_tokens/",
        "roles": "/api/v2/users/1/roles/",
        "organizations": "/api/v2/users/1/organizations/",
        "access_list": "/api/v2/users/1/access_list/",
        "teams": "/api/v2/users/1/teams/",
        "tokens": "/api/v2/users/1/tokens/",
        "personal_tokens": "/api/v2/users/1/personal_tokens/",
        "credentials": "/api/v2/users/1/credentials/",
        "activity_stream": "/api/v2/users/1/activity_stream/",
        "projects": "/api/v2/users/1/projects/"
    },
    "summary_fields": {
        "user_capabilities": {
            "edit": true,
            "delete": true
        }
    },
    "created": "2018-05-22T06:08:05.932739Z",
    "username": "admin",
    "first_name": "Kai",
    "last_name": "Ying",
    "email": "admin@kingnet.com",
    "is_superuser": true,
    "is_system_auditor": false,
    "ldap_dn": "",
    "external_account": null,
    "auth": []
}
```

Can you confirm the struct fields and ensure there is no fields left?

File: types.go [hunk 8]
```diff
@@ -156,11 +156,26 @@ type ProjectUpdate struct {
 
 // Project represents the awx api project.
 type Project struct {
```
Comment (Colstuwjx) [inline types.go]: The project struct is also not same to the awx rest api, could you help me clear it?

```
{
    "id": 4,
    "type": "project",
    "url": "/api/v2/projects/4/",
    "related": {
        "named_url": "/api/v2/projects/Demo Project++Default/",
        "created_by": "/api/v2/users/1/",
        "last_job": "/api/v2/project_updates/3757/",
        "notification_templates_error": "/api/v2/projects/4/notification_templates_error/",
        "notification_templates_success": "/api/v2/projects/4/notification_templates_success/",
        "object_roles": "/api/v2/projects/4/object_roles/",
        "notification_templates_any": "/api/v2/projects/4/notification_templates_any/",
        "copy": "/api/v2/projects/4/copy/",
        "project_updates": "/api/v2/projects/4/project_updates/",
        "update": "/api/v2/projects/4/update/",
        "access_list": "/api/v2/projects/4/access_list/",
        "teams": "/api/v2/projects/4/teams/",
        "scm_inventory_sources": "/api/v2/projects/4/scm_inventory_sources/",
        "inventory_files": "/api/v2/projects/4/inventories/",
        "schedules": "/api/v2/projects/4/schedules/",
        "playbooks": "/api/v2/projects/4/playbooks/",
        "activity_stream": "/api/v2/projects/4/activity_stream/",
        "organization": "/api/v2/organizations/1/",
        "last_update": "/api/v2/project_updates/3757/"
    },
    "summary_fields": {
        "last_job": {
            "id": 3757,
            "name": "Demo Project",
            "description": "",
            "finished": "2018-06-15T09:33:35.856828Z",
            "status": "successful",
            "failed": false
        },
        "last_update": {
            "id": 3757,
            "name": "Demo Project",
            "description": "",
            "status": "successful",
            "failed": false
        },
        "organization": {
            "id": 1,
            "name": "Default",
            "description": ""
        },
        "created_by": {
            "id": 1,
            "username": "admin",
            "first_name": "Kai",
            "last_name": "Ying"
        },
        "object_roles": {
            "admin_role": {
                "id": 13,
                "description": "Can manage all aspects of the project",
                "name": "Admin"
            },
            "use_role": {
                "id": 15,
                "description": "Can use the project in a job template",
                "name": "Use"
            },
            "update_role": {
                "id": 16,
                "description": "May update project or inventory or group using the configured source update system",
                "name": "Update"
            },
            "read_role": {
                "id": 14,
                "description": "May view settings for the project",
                "name": "Read"
            }
        },
        "user_capabilities": {
            "edit": true,
            "start": true,
            "copy": true,
            "schedule": true,
            "delete": true
        }
    },
    "created": "2018-05-22T06:08:07.848477Z",
    "modified": "2018-06-15T09:33:35.868448Z",
    "name": "Demo Project",
    "description": "",
    "local_path": "_4__demo_project",
    "scm_type": "git",
    "scm_url": "https://github.com/ansible/ansible-tower-samples",
    "scm_branch": "",
    "scm_clean": false,
    "scm_delete_on_update": false,
    "credential": null,
    "timeout": 0,
    "last_job_run": "2018-06-15T09:33:35.856828Z",
    "last_job_failed": false,
    "next_job_run": null,
    "status": "successful",
    "organization": 1,
    "scm_delete_on_next_update": false,
    "scm_update_on_launch": true,
    "scm_update_cache_timeout": 0,
    "scm_revision": "347e44fea036c94d5f60e544de006453ee5c71ad",
    "custom_virtualenv": null,
    "last_update_failed": false,
    "last_updated": "2018-06-15T09:33:35.856828Z"
}
```

File: types.go (hunk 7 above)
Comment (ramakantar) [reply types.go]: I just checked. Some elements in the "Related" structure are missing. Will fix it now.

File: types.go (hunk 8 above)
Comment (ramakantar) [reply types.go]: Sure. Some fields are missing in the structure. Will fix it now and test in my local repository.

File: types.go (hunk 7 above)
Comment (Colstuwjx) [reply types.go]: Not ONLY `related` fields, but also some struct fields, like `User`, `Ldap` field is not exists in ListUsers response, and some fields like `is_system_auditor `, `ldap_dn`, `external_account` etc

Feel free to add comments!

File: types.go (hunk 8 above)
Comment (Colstuwjx) [reply types.go]: Okay, thanks!

File: types.go (hunk 7 above)
Comment (ramakantar) [reply types.go]: Following the doc https://docs.ansible.com/ansible-tower/2.3.0/html/towerapi/users.html
is_system_auditor and external_account does not exist.

Comment (Colstuwjx) [reply types.go]: Check the new version document: https://docs.ansible.com/ansible-tower/3.2.4/html/towerapi/users.html for details.

Comment (ramakantar) [reply types.go]: Ah alright. my bad. Will fix it now.

**Comment by Colstuwjx** [comment]
Thanks.

**Pull request merged**